tools/mipgen/src/main.cpp

```diff
@@ -0,0 +1,240 @@
+                }
+                if (arg == "rgbm") {
+                    g_format = ImageEncoder::Format::RGBM;
+                    g_formatSpecified = true;
+                }
+                if (arg == "exr") {
+                    g_format = ImageEncoder::Format::EXR;
+                    g_formatSpecified = true;
+                }
+                if (arg == "psd") {
+                    g_format = ImageEncoder::Format::PSD;
+                    g_formatSpecified = true;
+                }
+                if (arg == "dds") {
+                    g_format = ImageEncoder::Format::DDS_LINEAR;
+                    g_formatSpecified = true;
+                }
+                break;
+            case 'c':
+                g_compression = arg;
+                break;
+        }
+    }
+
+    return optind;
+}
+
+int main(int argc, char* argv[]) {
+    int optionIndex = handleArguments(argc, argv);
+    int numArgs = argc - optionIndex;
+    if (numArgs < 2) {
+        printUsage(argv[0]);
+        return 1;
+    }
+    Path inputPath(argv[optionIndex++]);
+    string outputPattern(argv[optionIndex]);
+    if (!g_formatSpecified) {
+        constexpr bool forceLinear = true;
+        g_format = ImageEncoder::chooseFormat(outputPattern, forceLinear);
+    }
+
+    puts("Reading image...");
+    ifstream inputStream(inputPath.getPath(), ios::binary);
+    LinearImage sourceImage = ImageDecoder::decode(inputStream, inputPath.getPath());
+    if (!sourceImage.isValid()) {
+        cerr << "Unable to open image: " << inputPath.getPath() << endl;
+        exit(1);
+    }
+
+    puts("Generating miplevels...");
+    uint32_t count = getMipmapCount(sourceImage);
+    vector<LinearImage> miplevels(count);
+    generateMipmaps(sourceImage, g_filter, miplevels.data(), count);
+
+    puts("Writing image files to disk...");
+    char path[256];
+    uint32_t mip = 1; // start at 1 because 0 is the original image
+    for (auto image: miplevels) {
+        int result = snprintf(path, sizeof(path), outputPattern.c_str(), mip++);
+        if (result < 0 || result >= sizeof(path)) {
+            puts("Output pattern is too long.");
+            exit(1);
+        }
+        ofstream outputStream(path, ios::binary | ios::trunc);
+        if (!outputStream) {
+            cerr << "The output file cannot be opened: " << path << endl;
+        } else {
+            ImageEncoder::encode(outputStream, g_format, image, g_compression, path);
+            outputStream.close();
+            if (!outputStream) {
+                cerr << "An error occurred while writing the output file: " << path << endl;
+            }
+        }
+    }
+
+    puts("Generating mipmaps.html...");
+    char tag[256];
+    mip = 1;
+    const char* pattern = R"(<image src="%s" width="%dpx" height="%dpx">)";
+    const uint32_t width = sourceImage.getWidth();
+    const uint32_t height = sourceImage.getHeight();
+    ofstream html("mipmaps.html", ios::trunc);
```

Comment:
Could we make the generation of the HTML file optional behind a flag? It might be annoying to end up with a bunch of mipmaps.html all over our hard drives :)